File: python/mirage/observe/redis_store.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Awaitable
from typing import cast

import redis.asyncio as aioredis

from mirage.observe.store import ObserverStoreBase


class RedisObserverStore(ObserverStoreBase):
    """ObserverStore backed by Redis strings (one key per JSONL file).

    Appends use the atomic Redis APPEND command; an index set tracks
    which file keys exist so read_all needs no SCAN.

    Args:
        url (str): Redis connection URL.
        key_prefix (str): Namespace prefix for all keys.
    """

    def __init__(
        self,
        url: str = "redis://localhost:6379/0",
        key_prefix: str = "mirage:observer:",
    ) -> None:
        self._client = aioredis.from_url(url)
        self._prefix = key_prefix
        self._index_key = f"{key_prefix}keys"

    async def append(self, path: str, data: bytes) -> None:
        """Append bytes to the file at path.

        Args:
            path (str): File key.
            data (bytes): Bytes to append.
        """
        pipe = self._client.pipeline()
        pipe.append(f"{self._prefix}{path}", data)
        pipe.sadd(self._index_key, path)
        await pipe.execute()

    async def write(self, path: str, data: bytes) -> None:
        """Overwrite the file at path.

        Args:
            path (str): File key.
            data (bytes): Full new content.
        """
        pipe = self._client.pipeline()
        pipe.set(f"{self._prefix}{path}", data)
        pipe.sadd(self._index_key, path)
        await pipe.execute()

    async def read_matching(self, suffix: str) -> dict[str, bytes]:
        """Read only the files whose key ends with suffix.

        Filters on the index set, so only matching values transfer.

        Args:
            suffix (str): File-key suffix.

        Returns:
            dict[str, bytes]: Mapping of matching key to content.
        """
        paths = [p for p in await self._indexed_paths() if p.endswith(suffix)]
        return await self._read_paths(paths)

    async def _indexed_paths(self) -> list[str]:
        members = await cast("Awaitable[set[bytes]]",
                             self._client.smembers(self._index_key))
        return sorted(m.decode() if isinstance(m, bytes) else m
                      for m in members)

    async def _read_paths(self, paths: list[str]) -> dict[str, bytes]:
        if not paths:
            return {}
        values = await self._client.mget([f"{self._prefix}{p}" for p in paths])
        return {p: v or b"" for p, v in zip(paths, values)}

    async def clear(self) -> None:
        """Delete every stored file and the index (rewind/teardown)."""
        paths = await self._indexed_paths()
        keys = [f"{self._prefix}{p}" for p in paths] + [self._index_key]
        await self._client.delete(*keys)

    async def close(self) -> None:
        """Close the Redis client connection."""
        await self._client.aclose()

File: typescript/packages/node/src/observe/redis_store.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { RedisClientType } from 'redis'
import type { ObserverStore } from '@struktoai/mirage-core'
import { loadOptionalPeer } from '../optional_peer.ts'

export interface RedisObserverStoreOptions {
  url?: string
  keyPrefix?: string
}

/**
 * ObserverStore backed by Redis strings (one key per JSONL file).
 *
 * Appends use the atomic Redis APPEND command; an index set tracks which
 * file keys exist so readAll needs no SCAN. Mirrors the Python
 * RedisObserverStore.
 */
export class RedisObserverStore implements ObserverStore {
  readonly url: string
  private readonly prefix: string
  private readonly indexKey: string
  private clientPromise: Promise<RedisClientType> | null = null

  constructor(options: RedisObserverStoreOptions = {}) {
    this.url = options.url ?? 'redis://localhost:6379/0'
    this.prefix = options.keyPrefix ?? 'mirage:observer:'
    this.indexKey = `${this.prefix}keys`
  }

  private async client(): Promise<RedisClientType> {
    this.clientPromise ??= (async () => {
      const mod = await loadOptionalPeer(
        () =>
          import('redis') as unknown as Promise<{
            createClient: (o: { url: string }) => RedisClientType
          }>,
        { feature: 'RedisObserverStore', packageName: 'redis' },
      )
      const c = mod.createClient({
        url: this.url,
        socket: { reconnectStrategy: false },
      } as Parameters<typeof mod.createClient>[0])
      await c.connect()
      return c
    })()
    return this.clientPromise
  }

  async append(key: string, data: Uint8Array): Promise<void> {
    const c = await this.client()
    const buf = Buffer.from(data.buffer, data.byteOffset, data.byteLength)
    await c.multi().append(`${this.prefix}${key}`, buf).sAdd(this.indexKey, key).exec()
  }

  async write(key: string, data: Uint8Array): Promise<void> {
    const c = await this.client()
    const buf = Buffer.from(data.buffer, data.byteOffset, data.byteLength)
    await c.multi().set(`${this.prefix}${key}`, buf).sAdd(this.indexKey, key).exec()
  }

  readAll(): Promise<Map<string, Uint8Array>> {
    return this.readMatching('')
  }

  async readMatching(suffix: string): Promise<Map<string, Uint8Array>> {
    const paths = (await this.indexedPaths()).filter((p) => p.endsWith(suffix))
    return this.readPaths(paths)
  }

  private async indexedPaths(): Promise<string[]> {
    const c = await this.client()
    const members = await c.sMembers(this.indexKey)
    return [...members].sort()
  }

  private async readPaths(paths: string[]): Promise<Map<string, Uint8Array>> {
    const out = new Map<string, Uint8Array>()
    if (paths.length === 0) return out
    const c = await this.client()
    const mod = (await import('redis')) as unknown as {
      RESP_TYPES: { readonly BLOB_STRING: number }
    }
    const mapping: Record<number, unknown> = { [mod.RESP_TYPES.BLOB_STRING]: Buffer }
    const typed = c as unknown as {
      withTypeMapping: (m: Record<number, unknown>) => {
        mGet: (keys: string[]) => Promise<(Buffer | null)[]>
      }
    }
    const values = await typed.withTypeMapping(mapping).mGet(paths.map((p) => `${this.prefix}${p}`))
    paths.forEach((p, i) => {
      const raw = values[i]
      out.set(p, raw === null || raw === undefined ? new Uint8Array() : new Uint8Array(raw))
    })
    return out
  }

  async clear(): Promise<void> {
    const c = await this.client()
    const paths = await this.indexedPaths()
    const keys = [...paths.map((p) => `${this.prefix}${p}`), this.indexKey]
    if (keys.length > 0) await c.del(keys)
  }

  async close(): Promise<void> {
    // Idempotent: the workspace closes the plane store it consumed and the
    // owning WorkspaceStateStore closes every plane it built; the second
    // close must be a no-op, not a crash on an already-quit client.
    if (this.clientPromise === null) return
    const pending = this.clientPromise
    this.clientPromise = null
    const c = await pending
    await c.quit()
  }
}
